feat: Clarify Rootless Runtime Requirements (#4022)
kubernetes-sigs/kind: site/content/docs/user/rootless.md

Starting with kind 0.11.0, [Rootless Docker](https://docs.docker.com/go/rootless/), [Rootless Podman](https://github.com/containers/podman/blob/master/docs/tutorials/rootless_tutorial.md) and [Rootless nerdctl](https://github.com/containerd/nerdctl/blob/main/docs/rootless.md) can be used as the node provider of kind.

## Provider requirements

> 

Reviewer: ?

Author: Markdown style consistency - I like having an empty line after a heading. This doesn't appear to impact the site rendering by Hugo (see the deploy preview).

- Docker: 20.10 or later
- Podman: 3.0 or later
- nerdctl: 1.7 or later

## Host requirements

> ### cgroups v2

Reviewer: ```suggestion
### cgroup v2
```

For consistency with the other occurrences

The host needs to be running with cgroup v2.
Make sure that the result of the `docker info` command contains `Cgroup Version: 2`.
If it prints `Cgroup Version: 1`, try adding `GRUB_CMDLINE_LINUX="systemd.unified_cgroup_hierarchy=1"` to `/etc/default/grub` and
running `sudo update-grub` to enable cgroup v2.

Your host may also need to enable [cgroup delegation](https://systemd.io/CGROUP_DELEGATION/) for daemon-based controller runtimes.

> This is not required for daemonless runtimes, such as podman. Note that this procedure may

Reviewer: Untrue. What's the source of this misinformation? Is this from some hallucinating LLM?

[negatively impact performance](https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/ZMKLS7SHMRJLJ57NZCYPBAQ3UOYULV65/).

To enable cgroup delegation, perform the folowing actions:

1. As root, create the directory `/etc/systemd/system/user@.service.d/` if it does not already exist

   ```sh

>    sudo mdkir -p /etc/systemd/system/user@.service.d/

Reviewer: typo: mdkir

Reviewer: Is that obvious that mkdir is needed when the directory does not exist?

Author: Convenience for users - code blocks are easy to copy/paste.

Reviewer 2: Yeah, we will get a lot of novice users that are not familiar with any of this. A lot of Linux users will run less common configurations without much experience.

We occasionally get a bug from someone using a custom CPU scheduler that lacks cgroups .... etc.

   ```
2. As root, create the file `/etc/systemd/system/user@.service.d/delegate.conf` with the following content:

   ```ini
   [Service]
   Delegate=yes
   ```

3. Reload the systemd daemon:

   ```sh
   sudo systemctl daemon-reload
   ```

4. If using docker, reload the user docker daemon:

   ```sh
   systemctl --user restart docker
   ```

### Networking

> Containers running in rootless mode are not typically loaded with host-level iptable modules.

Author: My Linux experience is almost exclusively with Fedora. I am not familiar with what kernel mods are loaded by default in other distributions.

> This breaks the behavior of most Ingress and Gateway controllers.

Reviewer: Not specific to these controllers

Author: Better to phrase as "This breaks the behavior of many networking components, such as Ingress and Gateway controllers"?

> To load the iptable modules into the KinD containers, do the following:

Author: My question - why isn't this procedure needed for rootful runtimes? How are these kernel modules getting loaded dynamically?

As with the other modifications here, these are system-level changes that may have undesirable side effects. I think it's helpful if end users understand the "deeper why" behind these changes. For all I know, this is a feature gap that could be fixed somewhere in the stack.

> To load the iptable modules into the KinD containers, do the following:

Reviewer: Because the modules are loaded on demand in the case of rootful

## Provider requirements

- Docker: 20.10 or later
- Podman: 3.0 or later
- nerdctl: 1.7 or later

## Host requirements

### cgroup v2

The host needs to be running with cgroup v2, which is the default for many Linux disributions:

- Ubuntu: 22.04 and later.
- Fedora: 31 and later.
- Arch: April 2021 release and later.

You can verify the cgroup version used by your controller runtime with the following procedure:

- `docker`: Run `docker info` and look for `Cgroup Version: 2` in the output.
- `podman`: Run `podman info` and look for `cgroupVersion: v2` in the output.
- `nerdctl`: Run `nerdctl info` and look for `Cgroup Version: 2` in the output.

If the `info` output prints `Cgroup Version: 1` or equivalent, try the following to enable cgroup v2:

1. In `/etc/default/grub`, add the line `GRUB_CMDLINE_LINUX="systemd.unified_cgroup_hierarchy=1"`
2. Run `sudo update-grub` to enable cgroup v2.

Your host will also need to enable [cgroup delegation](https://systemd.io/CGROUP_DELEGATION/) for
user services. This is enabled by default for distributions running `systemd` version 252 and higher.

> To enable cgroup delegation, do the following:

Reviewer: ```suggestion
To enable cgroup delegation for all the controllers, do the following:
```

1. Check your version of `systemd` by running `systemctl --version`. If the output prints
   `systemd 252` or higher, no further action is needed. Example output below from a Fedora host:

   ```sh
   $ systemctl --version
   systemd 257 (257.9-2.fc42)
   ```

2. For systems with older versions of `systemd`, first create the directory
   `/etc/systemd/system/user@.service.d/` if it is not present.

   ```sh
   sudo mdkir -p /etc/systemd/system/user@.service.d/
   ```

3. Next, create the file `/etc/systemd/system/user@.service.d/delegate.conf` with the following content:

   ```ini
   [Service]
   Delegate=yes
   ```

4. Reload systemd for these changes to take effect:

   ```sh
   sudo systemctl daemon-reload
   ```

5. If using docker, reload the user docker daemon:

   ```sh
   systemctl --user restart docker
   ```

### Networking

Containers running in rootless mode may not loaded with host-level iptable modules.
This breaks the behavior of most networking components, such as Ingress and Gateway controllers.

> To load the iptable modules into the KIND containers, do the following:

Reviewer: ```suggestion
To load the iptable modules, do the following:
```

1. First, use `lsmod` to check which kernel modules are loaded by default for user processes on
   your system. Use `grep` to find which iptable modules are loaded:

   ```sh
   lsmod | grep "ip.*table"
   ```

2. Check the output for the following kernel modules:
   - `ip6_tables`
   - `ip6table_nat`
   - `ip_tables`
   - `iptable_nat`

3. If one or more of the kernel modules above are not present, your system needs to load these at
   startup for each process. First, run the following command to add these missing modules:
   
   ```sh
   sudo tee /etc/modules-load.d/iptables.conf > /dev/null <<'EOF'
   ip6_tables
   ip6table_nat
   ip_tables
   iptable_nat
   EOF
   ```

4. Check that the new module loading configuration is correct. You should see the following output:

   ```sh
   $ cat /etc/modules-load.d/iptables.conf 
   ip6_tables
   ip6table_nat
   ip_tables
   iptable_nat
   ```

> 4. Next, restart the `systemd-modules-load` service to make these changes effective immediately:

Reviewer 3: `4` is used twice. Most markdown renderers handle that fine, but would be good to be consistent by either making them all `1` to be obvious that the numbering is left to the rendering step, or to have them be sequentially numbered.

[particularly susceptible](https://github.com/kubernetes-sigs/kind/issues/3451) to this issue.

To increase the PID limit, do the following:

1. If using podman, edit your `containers.conf` file (generally located in
   `/etc/containers/containers.conf` or `~/.config/containers/containers.conf`) to increase the PIDs
   limit to a desired value (default 4096 on most systems):

    ```ini
    [containers]
    pids_limit = 65536
    ```

2. Re-recreate the KIND cluster for these changes to take effect:

   ```sh
   kind delete cluster && kind create cluster
   ```

### Increase inotify Limits

As documented in [known issues](/docs/user/known-issues/#pod-errors-due-to-too-many-open-files), pods may
fail by reaching inotify watch and instance limits. Ingress controllers such as NGINX and Contour
are particularly susceptible to this issue.

To increase the inotify limits, do the following:

1. As root, create a `.conf` file in `/etc/systctl.d` that increases the `fs.inotify` max user settings:

   ```
   fs.inotify.max_user_watches = 524288
   fs.inotify.max_user_instances = 512
   ```

2. Reload `sysctl` for these changes to take effect:

   ```sh
   sudo sysctl --system
   ```

> 3. Alternatively, restart your system for these changes to take effect.

Reviewer: ```suggestion
Alternatively, restart your system for these changes to take effect.
```

> 

Reviewer 3: nit: If we're going to be picky about having a blank line after a header, we should also be picky about not having multiple blank lines. ;)

### Allow Binding to Privileged Ports

If you use the `extraPortMappings` method to provide ingress to your KIND cluster, you can allow
the KIND node container to bind to ports 80 and 443 on the host. User containers cannot bind to

> these ports by default as they are considered privileged.

Reviewer: ```suggestion
ports below 1024 by default as they are considered privileged.
```

  iptable_nat
  ```

> 2. Restart your system to ensure these changes take effect.

Reviewer: `sudo systemctl restart systemd-modules-load.service` should suffice

[process id limits](https://docs.podman.io/en/v4.3/markdown/options/pids-limit.html#pids-limit-limit)
that may be too low for the node or for a pod running on the node. The NGINX ingress controller is
[particularly susceptible](https://github.com/kubernetes-sigs/kind/issues/3451) to this issue.

To increase the PID limit, do the following:

1. If using podman, edit your `containers.conf` file (generally located in
   `/etc/containers/containers.conf` or `~/.config/containers/containers.conf`) to increase the PIDs
   limit to a desired value (default 4096 on most systems).

    ```ini
    [containers]
    pids_limit = 65536
    ```

### Increase inotify Limits

As documented in [known issues](/docs/user/known-issues/#pod-errors-due-to-too-many-open-files), pods may
fail by reaching inotify watch and instance limits. Ingress controllers such as NGINX and Contour
are particularly susceptible to this issue.

To increase the inotify limits, do the following:

1. As root, create a `.conf` file in `/etc/systctl.d` that increases the `fs.inotify` max user settings:

   ```
   fs.inotify.max_user_watches = 524288
   fs.inotify.max_user_instances = 512
   ```

> 2. Restart your system for these changes to take effect.

Reviewer: Why not `sysctl --system` ?

Author: Will add in a follow-up commit.

> ### Allow Unprivileged Binding to HTTP(S) Ports

Reviewer: Not specific to HTTP(S)
